src/f32-spmm/gen/f32-spmm-128x1-minmax-hvx-pipelined-x2.c

```diff
@@ -0,0 +1,396 @@
+
+#include "xnnpack/simd/f32-hvx.h"
+
+#include "xnnpack/prefetch.h"
+#include "xnnpack/spmm.h"
+
+void xnn_f32_spmm_minmax_ukernel_128x1__hvx_pipelined_x2(
+    size_t mc,
+    size_t nc,
+    const float* input,
+    const float* weights,
+    const int32_t* widx_dmap,
+    const uint32_t* nidx_nnzmap,
+    float* output,
+    size_t output_stride,
+    const union xnn_f32_minmax_params params[restrict XNN_MIN_ELEMENTS(1)])
+{
+  assert(mc != 0);
+  assert(mc % sizeof(float) == 0);
+  assert(nc != 0);
+
+  const HVX_Vector vmin = xnn_set1_f32(params->scalar.min);
+  const HVX_Vector vmax = xnn_set1_f32(params->scalar.max);
+
+  size_t output_decrement = output_stride * nc - 128 * sizeof(float);
+  while XNN_LIKELY(mc >= 128 * sizeof(float)) {
+    const float* w = weights;
+    const int32_t* dmap = widx_dmap;
+    const uint32_t* nnzmap = nidx_nnzmap;
+    HVX_Vector vw = xnn_set1_f32(*w); w += 1;
+
+    intptr_t diff = *dmap++;
+
+    HVX_Vector vi0 = xnn_loadu_f32(input);
+    HVX_Vector vi1 = xnn_loadu_f32(input + 32);
+    HVX_Vector vi2 = xnn_loadu_f32(input + 64);
+    HVX_Vector vi3 = xnn_loadu_f32(input + 96);
+
+    size_t n = nc;
+
+    do {
+      uint32_t nnz = *nnzmap++;
+      HVX_Vector vacc0 = vw;
+      HVX_Vector vacc1 = vw;
+      HVX_Vector vacc2 = vw;
+      HVX_Vector vacc3 = vw;
+      vw = xnn_set1_f32(*w); w += 1;
+
+      for(; nnz >= 2; nnz -= 2) {
+        vacc0 = xnn_fmadd_f32(vi0, vw, vacc0);
+        vacc1 = xnn_fmadd_f32(vi1, vw, vacc1);
+        vacc2 = xnn_fmadd_f32(vi2, vw, vacc2);
+        vacc3 = xnn_fmadd_f32(vi3, vw, vacc3);
+        input = (const float*) ((uintptr_t) input + (uintptr_t) diff);
+
+        // Use `xnn_prefetch_to_l2` first to prefetch data into the L2 cache,
+        // then, use `xnn_prefetch_to_l1` to bring data into the L1 cache.
+        // format: xnn_prefetch_to_l1(addr)
+        //         xnn_prefetch_to_l2(addr, stride, width, height, iter)
+        xnn_prefetch_to_l2_1d(input + 32, 128);
+        xnn_prefetch_to_l1(input + 32);
+
+        diff = *dmap++;
+
+        vw = xnn_set1_f32(*w); w += 1;
+        xnn_prefetch_to_l2_1d(w + 32, 32);
+        xnn_prefetch_to_l1(w + 32);
+
+        vi0 = xnn_loadu_f32(input + 0);
+        vi1 = xnn_loadu_f32(input + 32);
+        vi2 = xnn_loadu_f32(input + 64);
+        vi3 = xnn_loadu_f32(input + 96);
+        vacc0 = xnn_fmadd_f32(vi0, vw, vacc0);
+        vacc1 = xnn_fmadd_f32(vi1, vw, vacc1);
+        vacc2 = xnn_fmadd_f32(vi2, vw, vacc2);
+        vacc3 = xnn_fmadd_f32(vi3, vw, vacc3);
+        input = (const float*) ((uintptr_t) input + (uintptr_t) diff);
+
+        // Use `xnn_prefetch_to_l2` first to prefetch data into the L2 cache,
+        // then, use `xnn_prefetch_to_l1` to bring data into the L1 cache.
+        // format: xnn_prefetch_to_l1(addr)
+        //         xnn_prefetch_to_l2(addr, stride, width, height, iter)
+        xnn_prefetch_to_l2_1d(input + 32, 128);
+        xnn_prefetch_to_l1(input + 32);
```

Comment:
I don't think prefetching vector data to L1 is useful. If nothing else, this is only fetching one cache line of 8(?) that are needed. But since vector data only reaches L2, I don't think we should do it at all.

Reply:
I will leave l2 prefetch only for now.